File: public/benchmarks/simd/zig/simd_vec.zig
const std = @import("std");
const N = 20000000;
var a: [N]f64 = undefined;
pub fn main() void {
    var i: usize = 0;
    while (i < N) : (i += 1) a[i] = @as(f64, @floatFromInt(i % 1000)) * 0.001;
    var acc: @Vector(4, f64) = @splat(0.0);
    const limit = (N / 4) * 4;
    i = 0;
    while (i < limit) : (i += 4) {
        const v: @Vector(4, f64) = a[i..][0..4].*;
        acc += v * v;
    }
    var sum: f64 = @reduce(.Add, acc);
    while (i < N) : (i += 1) sum += a[i] * a[i];
    std.debug.print("{d:.6}\n", .{sum});
}

File: public/benchmarks/simd/rust/simd_vec.rs
use std::arch::aarch64::*;

const N: usize = 20_000_000;

fn main() {
    let mut a = vec![0.0f64; N];
    for i in 0..N {
        a[i] = (i % 1000) as f64 * 0.001;
    }

    let limit = (N / 2) * 2;
    let mut sum: f64;
    unsafe {
        let mut acc = vdupq_n_f64(0.0);
        let mut i = 0;
        while i < limit {
            let v = vld1q_f64(a.as_ptr().add(i));
            acc = vfmaq_f64(acc, v, v);
            i += 2;
        }
        sum = vaddvq_f64(acc);
        while i < N {
            sum += a[i] * a[i];
            i += 1;
        }
    }
    println!("{:.6}", sum);
}
